File: python/modcdp/injector/BorrowedExtensionInjector.py
from __future__ import annotations

import time
from pathlib import Path
from typing import Any, Mapping, cast

from ..injector.ExtensionInjector import DEFAULT_SERVICE_WORKER_POLL_INTERVAL_MS, DEFAULT_SERVICE_WORKER_PROBE_TIMEOUT_MS, DEFAULT_SERVICE_WORKER_READY_TIMEOUT_MS, EXT_ID_FROM_URL_RE, ExtensionInjector, ExtensionInjectionResult, MODCDP_READY_EXPRESSION


class BorrowedExtensionInjector(ExtensionInjector):
    def inject(self) -> ExtensionInjectionResult | None:
        deadline = time.monotonic() + (self.options.get("injector_service_worker_ready_timeout_ms") or DEFAULT_SERVICE_WORKER_READY_TIMEOUT_MS) / 1000
        while True:
            borrowed = self._borrowVisibleServiceWorkers()
            if borrowed:
                return borrowed
            if time.monotonic() >= deadline:
                return None
            time.sleep((self.options.get("injector_service_worker_poll_interval_ms") or DEFAULT_SERVICE_WORKER_POLL_INTERVAL_MS) / 1000)

    def _borrowVisibleServiceWorkers(self) -> ExtensionInjectionResult | None:
        borrowed: list[ExtensionInjectionResult] = []
        visible_service_workers = [
            target
            for target in self._targetInfos()
            if target.get("type") == "service_worker" and isinstance(target.get("url"), str) and target["url"].startswith("chrome-extension://")
        ]
        has_configured_matcher = bool(
            self.options.get("injector_extension_id")
            or self.options.get("injector_service_worker_url_includes")
            or self.options.get("injector_service_worker_url_suffixes")
        )
        candidates = [target for target in visible_service_workers if self._serviceWorkerTargetMatches(target)] if has_configured_matcher else visible_service_workers
        for target in candidates:
            try:
                bootstrapped = self._bootstrapTarget(target)
            except Exception:
                bootstrapped = None
            if bootstrapped:
                borrowed.append({**bootstrapped, "source": "borrowed"})
        borrowed.sort(key=lambda item: (bool(item.get("has_debugger")), bool(item.get("has_tabs"))), reverse=True)
        return borrowed[0] if borrowed else None

    def _bootstrapTarget(self, target) -> ExtensionInjectionResult | None:
        session_id = self._ensureSessionIdForTarget(
            target["targetId"],
            self.options.get("injector_service_worker_probe_timeout_ms") or DEFAULT_SERVICE_WORKER_PROBE_TIMEOUT_MS,
            True,
        )
        if session_id is None:
            return None
        try:
            self._sendWithTimeout("Runtime.enable", {}, session_id)
        except Exception:
            pass
        bootstrap = self._sendWithTimeout(
            "Runtime.evaluate",
            {
                "expression": f"({bootstrap_modcdp_server_expression()})()",
                "awaitPromise": True,
                "returnByValue": True,
            },
            session_id,
        )
        result = cast(Mapping[str, Any], bootstrap.get("result")) if isinstance(bootstrap.get("result"), Mapping) else {}
        raw_value = result.get("value")
        value = cast(Mapping[str, Any], raw_value) if isinstance(raw_value, Mapping) else {}
        if not bool(value.get("has_tabs")) or not bool(value.get("has_debugger")):
            return None
        ready = bool(value.get("ok"))
        if ready and self._readyExpression() != MODCDP_READY_EXPRESSION:
            probe = self._sendWithTimeout(
                "Runtime.evaluate",
                {
                    "expression": self._readyExpression(),
                    "returnByValue": True,
                },
                session_id,
            )
            probe_result = cast(Mapping[str, Any], probe.get("result")) if isinstance(probe.get("result"), Mapping) else {}
            ready = bool(probe_result.get("value"))
        if not ready:
            return None
        match = EXT_ID_FROM_URL_RE.match(target["url"])
        extension_id = value.get("extension_id") if isinstance(value.get("extension_id"), str) else None
        return {
            "source": "borrowed",
            "extension_id": extension_id or (match.group(1) if match else None),
            "target_id": target["targetId"],
            "url": target["url"],
            "session_id": session_id,
            "has_tabs": bool(value.get("has_tabs")),
            "has_debugger": bool(value.get("has_debugger")),
        }


def bootstrap_modcdp_server_expression() -> str:
    source = modcdp_server_source()
    start = source.index("export function installModCDPServer")
    end = source.index("export const ModCDPServer")
    installer = source[start:end].replace("export function", "function", 1)
    return (
        "function() {\n"
        "const __name = (fn) => fn;\n"
        f"{installer}\n"
        "const ModCDP = installModCDPServer(globalThis);\n"
        "return {\n"
        "  ok: Boolean(ModCDP?.__ModCDPServerVersion >= 1 && ModCDP?.handleCommand && ModCDP?.addCustomEvent),\n"
        "  extension_id: globalThis.chrome?.runtime?.id ?? null,\n"
        "  has_tabs: Boolean(globalThis.chrome?.tabs?.query),\n"
        "  has_debugger: Boolean(globalThis.chrome?.debugger?.sendCommand && globalThis.chrome?.debugger?.getTargets),\n"
        "};\n"
        "}"
    )


def modcdp_server_source() -> str:
    candidates: list[Path] = []
    for parent in Path(__file__).resolve().parents:
        candidates.append(parent / "dist" / "js" / "src" / "server" / "ModCDPServer.js")
        candidates.append(parent / "dist" / "extension" / "js" / "src" / "server" / "ModCDPServer.js")
        candidates.append(parent / "dist" / "extension" / "ModCDPServer.js")
        candidates.append(parent / "ModCDPServer.js")
    for candidate in candidates:
        if candidate.exists():
            return candidate.read_text()
    checked = ", ".join(str(candidate) for candidate in candidates)
    raise FileNotFoundError(f"Unable to locate ModCDPServer.js; checked: {checked}")

File: go/modcdp/injector/BorrowedExtensionInjector.go
package injector

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"sort"
	"strings"
	"time"
)

type BorrowedExtensionInjector struct {
	ExtensionInjector
}

func NewBorrowedExtensionInjector(options ExtensionInjectorConfig) BorrowedExtensionInjector {
	return BorrowedExtensionInjector{ExtensionInjector: NewExtensionInjector(options)}
}

func (i *BorrowedExtensionInjector) Inject() (*ExtensionInjectionResult, error) {
	deadline := time.Now().Add(time.Duration(i.Options.InjectorServiceWorkerReadyTimeoutMS) * time.Millisecond)
	for {
		borrowed, err := i.borrowVisibleServiceWorkers()
		if err != nil || borrowed != nil {
			return borrowed, err
		}
		if time.Now().After(deadline) {
			return nil, nil
		}
		time.Sleep(time.Duration(i.Options.InjectorServiceWorkerPollIntervalMS) * time.Millisecond)
	}
}

func (i *BorrowedExtensionInjector) borrowVisibleServiceWorkers() (*ExtensionInjectionResult, error) {
	targets, err := i.targetInfos()
	if err != nil {
		return nil, err
	}
	hasConfiguredMatcher := i.Options.InjectorExtensionID != "" || len(i.Options.InjectorServiceWorkerURLIncludes) > 0 || len(i.Options.InjectorServiceWorkerURLSuffixes) > 0
	candidates := []map[string]any{}
	for _, target := range targets {
		targetType, _ := target["type"].(string)
		targetURL, _ := target["url"].(string)
		if targetType != "service_worker" || !strings.HasPrefix(targetURL, "chrome-extension://") {
			continue
		}
		if hasConfiguredMatcher && !i.serviceWorkerTargetMatches(target) {
			continue
		}
		candidates = append(candidates, target)
	}
	var borrowed []*ExtensionInjectionResult
	for _, target := range candidates {
		bootstrapped, err := i.bootstrapTarget(target)
		if err == nil && bootstrapped != nil {
			bootstrapped.Source = "borrowed"
			borrowed = append(borrowed, bootstrapped)
		}
	}
	sort.SliceStable(borrowed, func(left, right int) bool {
		if borrowed[left].HasDebugger != borrowed[right].HasDebugger {
			return borrowed[left].HasDebugger
		}
		return borrowed[left].HasTabs && !borrowed[right].HasTabs
	})
	if len(borrowed) == 0 {
		return nil, nil
	}
	return borrowed[0], nil
}

func (i *BorrowedExtensionInjector) bootstrapTarget(target map[string]any) (*ExtensionInjectionResult, error) {
	targetID, _ := target["targetId"].(string)
	targetURL, _ := target["url"].(string)
	sessionID := i.ensureSessionIDForTarget(targetID, i.Options.InjectorServiceWorkerProbeTimeoutMS, true)
	if sessionID == "" {
		return nil, nil
	}
	_, _ = i.sendWithTimeout("Runtime.enable", map[string]any{}, sessionID, i.Options.InjectorCDPSendTimeoutMS)
	bootstrap, err := modcdpServerBootstrapExpressionFromPath(i.Options.InjectorExtensionPath)
	if err != nil {
		return nil, err
	}
	probe, err := i.sendWithTimeout("Runtime.evaluate", map[string]any{
		"expression":    fmt.Sprintf("(%s)()", bootstrap),
		"awaitPromise":  true,
		"returnByValue": true,
	}, sessionID, i.Options.InjectorCDPSendTimeoutMS)
	if err != nil {
		return nil, err
	}
	result, _ := probe["result"].(map[string]any)
	value, _ := result["value"].(map[string]any)
	if hasTabs, _ := value["has_tabs"].(bool); !hasTabs {
		return nil, nil
	}
	if hasDebugger, _ := value["has_debugger"].(bool); !hasDebugger {
		return nil, nil
	}
	ready, _ := value["ok"].(bool)
	if ready && i.readyExpression() != modcdpReadyExpression {
		readyProbe, err := i.sendWithTimeout("Runtime.evaluate", map[string]any{
			"expression":    i.readyExpression(),
			"returnByValue": true,
		}, sessionID, i.Options.InjectorCDPSendTimeoutMS)
		if err != nil {
			return nil, err
		}
		readyResult, _ := readyProbe["result"].(map[string]any)
		ready, _ = readyResult["value"].(bool)
	}
	if !ready {
		return nil, nil
	}
	extensionID, _ := value["extension_id"].(string)
	if extensionID == "" {
		if match := extIDFromURL.FindStringSubmatch(targetURL); len(match) > 1 {
			extensionID = match[1]
		}
	}
	hasTabs, _ := value["has_tabs"].(bool)
	hasDebugger, _ := value["has_debugger"].(bool)
	return &ExtensionInjectionResult{
		Source:      "borrowed",
		ExtensionID: extensionID,
		TargetID:    targetID,
		URL:         targetURL,
		SessionID:   sessionID,
		HasTabs:     hasTabs,
		HasDebugger: hasDebugger,
	}, nil
}

func modcdpServerBootstrapExpressionFromPath(extensionPath string) (string, error) {
	serverPath, err := modcdpServerPathFromExtensionPath(extensionPath)
	if err != nil {
		return "", err
	}
	body, err := os.ReadFile(serverPath)
	if err != nil {
		return "", err
	}
	source := string(body)
	start := strings.Index(source, "export function installModCDPServer")
	end := strings.Index(source, "export const ModCDPServer")
	if start < 0 || end < start {
		return "", fmt.Errorf("could not find installModCDPServer in ModCDPServer.js")
	}
	installer := strings.Replace(source[start:end], "export function", "function", 1)
	return fmt.Sprintf(`function() {
const __name = (fn) => fn;
%s
const ModCDP = installModCDPServer(globalThis);
return {
  ok: Boolean(ModCDP?.__ModCDPServerVersion >= 1 && ModCDP?.handleCommand && ModCDP?.addCustomEvent),
  extension_id: globalThis.chrome?.runtime?.id ?? null,
  has_tabs: Boolean(globalThis.chrome?.tabs?.query),
  has_debugger: Boolean(globalThis.chrome?.debugger?.sendCommand && globalThis.chrome?.debugger?.getTargets),
};
}`, installer), nil
}

func modcdpServerPathFromExtensionPath(extensionPath string) (string, error) {
	candidates := []string{}
	if extensionPath != "" {
		candidates = append(candidates, filepath.Join(extensionPath, "ModCDPServer.js"))
		candidates = append(candidates, filepath.Join(extensionPath, "js", "src", "server", "ModCDPServer.js"))
		candidates = append(candidates, filepath.Join(filepath.Dir(extensionPath), "js", "src", "server", "ModCDPServer.js"))
	}
	if _, file, _, ok := runtime.Caller(0); ok {
		for dir := filepath.Dir(file); ; dir = filepath.Dir(dir) {
			candidates = append(candidates, filepath.Join(dir, "dist", "js", "src", "server", "ModCDPServer.js"))
			candidates = append(candidates, filepath.Join(dir, "dist", "extension", "js", "src", "server", "ModCDPServer.js"))
			candidates = append(candidates, filepath.Join(dir, "dist", "extension", "ModCDPServer.js"))
			if parent := filepath.Dir(dir); parent == dir {
				break
			}
		}
	}
	for _, candidate := range candidates {
		if _, err := os.Stat(candidate); err == nil {
			return candidate, nil
		}
	}
	return "", fmt.Errorf("unable to locate ModCDPServer.js; checked: %s", strings.Join(candidates, ", "))
}
